codeflash/code_utils/oauth_handler.py

```diff
@@ -0,0 +1,731 @@
+        code_challenge = base64.urlsafe_b64encode(hashlib.sha256(code_verifier.encode()).digest()).rstrip(b"=").decode()
+        return code_verifier, code_challenge
+
+    def start_local_server(self, port: int) -> http.server.HTTPServer:
+        """Start local HTTP server for OAuth callback."""
+        handler_class = self.create_callback_handler()
+        httpd = http.server.HTTPServer(("localhost", port), handler_class)
+
+        def serve_forever_wrapper() -> None:
+            httpd.serve_forever()
+
+        server_thread = threading.Thread(target=serve_forever_wrapper)
+        server_thread.daemon = True
+        server_thread.start()
+
+        return httpd
+
+    def wait_for_callback(self, httpd: http.server.HTTPServer, timeout: int = 120) -> bool:  # noqa: ARG002
+        """Wait for OAuth callback with timeout."""
+        waited = 0
+        while not self.is_complete and waited < timeout:
+            time.sleep(0.5)
+            waited += 0.5
+
+        return self.is_complete
+
+    def exchange_code_for_token(self, code: str, code_verifier: str, redirect_uri: str) -> str | None:
+        """Exchange authorization code for API token."""
+        token_url = f"{get_cfapi_base_urls().cfwebapp_base_url}/codeflash/auth/oauth/token"
+        data = {
+            "grant_type": "authorization_code",
+            "code": code,
+            "code_verifier": code_verifier,
+            "redirect_uri": redirect_uri,
+            "client_id": "cf_vscode_app",
+        }
+
+        try:
+            resp = requests.post(
+                token_url, headers={"Content-Type": "application/json"}, data=json.dumps(data), timeout=10
+            )
+            resp.raise_for_status()
+            token_json = resp.json()
+            api_key = token_json.get("access_token")
+
+            if not api_key:
+                self.token_error = "No access token in response"  # noqa: S105
+                return None
+
+        except requests.exceptions.HTTPError as e:
+            error_msg = f"HTTP {e.response.status_code}"
+            try:
+                error_data = e.response.json()
+                error_msg = error_data.get("error_description", error_data.get("error", error_msg))
+            except Exception:  # noqa: S110
+                pass
+            self.token_error = "Unauthorized"  # noqa: S105
+            click.echo(f"❌ {self.token_error}")
+            return None
+        except Exception:
+            self.token_error = "Unauthorized"  # noqa: S105
+            click.echo(f"❌ {self.token_error}")
+            return None
+        else:
+            return api_key
+
+
+def perform_oauth_signin() -> str | None:
+    """Perform OAuth PKCE flow and return API key if successful.
+
+    Returns None if failed.
+    """
+    oauth = OAuthHandler()
+
+    # Setup PKCE
+    port = oauth.get_free_port()
+    redirect_uri = f"http://localhost:{port}/callback"
+    code_verifier, code_challenge = oauth.generate_pkce_pair()
```

Comment:
is generating this in the cli side where the user can modify this generation logic a security vulnerability? what is user has full control of the generation string, would that degrade the security somehow?